### PR TITLE
CMake build extensions and IDL compiler change supporting IDL files including files in subdirectories

### PR DESCRIPTION
**Pull request by jonesc-oci**

### CODE REVIEW

File: tests/cmake_integration/idl_test/ecu.mpc [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+	ecu.idl
+  }
+
+  Idl_Files {
+    idlflags += -I./engine -I./engine/engine_stats -I./transmission
+	engine/engine_specs.idl
+	engine/engine_stats/fuel_stats.idl
+	transmission/transmission_specs.idl
+	transmission/transmission_states.idl
+	ecu.idl
```
Comment (iguessthislldo) [inline tests/cmake_integration/idl_test/ecu.mpc:23]: Tabs

File: tests/cmake_integration/idl_test/CMakeLists.txt [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+set(IDL_DEPENDENCIES ./engine/engine_stats/fuel_stats.idl)
+
+# add idl given a path. this updates IDL_DEPENDENCIES.
+ADD_IDL_PATH(./engine)
+ADD_IDL_FILES_IN_PATH(./engine)
+
+# add idl given a path. this updates IDL_DEPENDENCIES.
+ADD_IDL_PATH(./transmission)
+# this will import both transmission idl files
+ADD_IDL_FILES_IN_PATH(./transmission)
```
Comment (iguessthislldo) [inline tests/cmake_integration/idl_test/CMakeLists.txt:56]: I honestly don't like any of this looks. I would perfer any solution to this problem being invisible to the user if possible or at the very least contained within the call to `OPENDDS_TARGET_SOURCES`.

File: cmake/api_macros.cmake [hunk 3]
```diff
@@ -167,3 +167,20 @@ macro(OPENDDS_TARGET_SOURCES target)
 
   endforeach()
 endmacro()
+
+macro(ADD_DDS_IDL_FLAG flag)
```
Comment (iguessthislldo) [inline cmake/api_macros.cmake:171]: These names are in the global scope, so they should be prefixed with `OPENDDS_`.

File: cmake/api_macros.cmake [hunk 4]
```diff
@@ -167,3 +167,20 @@ macro(OPENDDS_TARGET_SOURCES target)
...
+  list(APPEND EXTRA_IDL_FLAGS "${flag}")
```
Comment (iguessthislldo) [inline cmake/api_macros.cmake:172]: Same thing with variables

**Review by iguessthislldo — CHANGES_REQUESTED** [review]

**Comment by iguessthislldo** [comment]
I forgot to add to my review that directories with CMake tests have to added to [this file directly or indirectly](https://github.com/objectcomputing/OpenDDS/blob/master/tests/cmake_integration/CMakeLists.txt), otherwise they won't be built. CMake doesn't have a recursive mechanism for adding subprojects like MPC does.

File: tests/cmake_integration/idl_test/.gitignore [hunk 5]
```diff
@@ -0,0 +1,69 @@
...
+fuel_statsTypeSupportC.cpp
+fuel_statsTypeSupportC.h
+fuel_statsTypeSupportC.inl
+fuel_statsTypeSupport.idl
+fuel_statsTypeSupportImpl.cpp
+fuel_statsTypeSupportImpl.h
+fuel_statsTypeSupportS.cpp
+fuel_statsTypeSupportS.h
+GNUmakefile
+GNUmakefile.ECU
```
Comment (iguessthislldo) [inline tests/cmake_integration/idl_test/.gitignore:43]: These should be covered by the general gitignore rules.

```suggestion
```

File: tests/cmake_integration/idl_test/CMakeLists.txt [hunk 6]
```diff
@@ -0,0 +1,52 @@
...
+
+set(CMAKE_CXX_COMPILER ${OPENDDS_COMPILER})
+
+set(opendds_libs
+  OpenDDS::Dcps # Core OpenDDS Library
+  OpenDDS::Rtps OpenDDS::Rtps_Udp # For run_test.pl --rtps
+)
+
+## debugging
+set(CMAKE_VERBOSE_MAKEFILE on)
```
Comment (iguessthislldo) [inline tests/cmake_integration/idl_test/CMakeLists.txt:17]: Does this need to be left on?

File: tests/cmake_integration/idl_test/CMakeLists.txt [hunk 7]
```diff
@@ -0,0 +1,52 @@
...
+project(OpenDDS_car_cmake CXX)
+cmake_minimum_required(VERSION 3.8.2)
+
+find_package(OpenDDS REQUIRED)
+
+set(CMAKE_CXX_COMPILER ${OPENDDS_COMPILER})
+
+set(opendds_libs
+  OpenDDS::Dcps # Core OpenDDS Library
+  OpenDDS::Rtps OpenDDS::Rtps_Udp # For run_test.pl --rtps
```
Comment (iguessthislldo) [inline tests/cmake_integration/idl_test/CMakeLists.txt:13]:
```suggestion
```

**Review by iguessthislldo — CHANGES_REQUESTED** [review]
One last thing is the name of the test. Could you make it a little more specific than "idl_test"?

File: cmake/init.cmake [hunk 8]
```diff
@@ -40,7 +40,7 @@ if(NOT DEFINED DDS_ROOT)
     set(OPENDDS_LIB_DIR "${DDS_ROOT}/lib")
   endif()
 
-else()
+elseif(NOT DEFINED DDS_ALLOW_ENV_REDEFINES)
```
Comment (iguessthislldo) [inline cmake/init.cmake:43]: This should start with `OPENDDS` and should probably this should probably be an [option/cache variable](https://github.com/objectcomputing/OpenDDS/blob/26d4b1dfc663c1347169a4ca3082f96a04c309d9/cmake/init.cmake#L11-L12). You could also make the include filename only thing a global setting if you want.

Can you also document this in and everything else [`cmake.md`](https://github.com/objectcomputing/OpenDDS/blob/master/docs/cmake.md#cache-variablesoptions-understood-by-opendds)?

**Review by iguessthislldo — APPROVED** [review]

File: cmake/init.cmake (hunk 8 above)
Comment (iguessthislldo) [reply cmake/init.cmake:43]: I accidentally clicked approved and I can't seem to change it.

File: tests/cmake_integration/Nested_IDL/CMakeLists_duplicate_filename.txt [hunk 9]
```diff
@@ -0,0 +1,50 @@
+# Distributed under the OpenDDS License. See accompanying LICENSE
```
Comment (iguessthislldo) [inline tests/cmake_integration/Nested_IDL/CMakeLists_duplicate_filename.txt:1]: I don't think this file is going to be picked up by CMake like this is set up now.

**Review by iguessthislldo — CHANGES_REQUESTED** [review]

File: tests/cmake_integration/Nested_IDL/CMakeLists_duplicate_filename.txt (hunk 9 above)
Comment (jonesc-oci) [reply tests/cmake_integration/Nested_IDL/CMakeLists_duplicate_filename.txt:1]: I'll remove this file.

File: cmake/api_macros.cmake [hunk 10]
```diff
@@ -92,13 +92,54 @@ macro(OPENDDS_TARGET_SOURCES target)
...
+  list(LENGTH the_idl_files idl_file_count)
+  list(REMOVE_DUPLICATES the_idl_files)
+  list(LENGTH the_idl_files unique_idl_file_count)
+
+  if(NOT ${idl_file_count} EQUAL ${unique_idl_file_count})
+    message(WARNING "OPENDDS_TARGET_SOURCES has 2 or more IDL files with \
+    the same name. ${check_for_dups_list}")
+  endif()
+
+  if(NOT "NO_FILENAME_ONLY_INCLUDES" IN_LIST arglist)
```
[the remark this answers is not in the document]
Comment (jonesc-oci) [reply cmake/api_macros.cmake:118]: I'll make it an option rather than a flag to the macro.

**Review by iguessthislldo — APPROVED** [review]

**Comment by iguessthislldo** [comment]
```
ERROR: dds/idl/be_global.cpp:558 failed tabs check
ERROR: dds/idl/be_global.cpp:560 failed tabs check
ERROR: dds/idl/be_global.cpp:561 failed tabs check
ERROR: dds/idl/be_global.cpp:563 failed tabs check
ERROR: dds/idl/be_global.cpp:564 failed tabs check
ERROR: dds/idl/be_global.cpp:565 failed tabs check
ERROR: dds/idl/be_global.cpp:567 failed tabs check
```

**Pull request merged**